src/sfptpd_clock.c

```diff
@@ -2484,4 +2571,163 @@ bool sfptpd_clock_is_active(const struct sfptpd_clock *clock)
+	const useconds_t settle = 1000000;
+	struct sfptpd_timespec test_adjustment;
+	struct sfptpd_timespec restore_adjustment;
+	struct sfptpd_timespec threshold;
+	struct sfptpd_timespec t;
+	struct sfptpd_clock_calibration *calibrations;
+	int i, j;
+	int rc;
+	int changes = 0;
+
+	clock_lock();
+
+	snapshot = clock_snapshot(&count, true);
+
+	if (count < 2)
+		goto finish;
+
+	NOTICE("clock: identifying duplicate phc clocks, a destructive operation\n");
+
+	sys = sfptpd_clock_get_system_clock();
+	sfptpd_time_init(&test_adjustment, -1L, 0U, 0U);
+	sfptpd_time_init(&threshold, 0L, 200000000U, 0U);
+	sfptpd_time_negate(&restore_adjustment, &test_adjustment);
+	calibrations = calloc(count, sizeof *calibrations);
+	if (calibrations == NULL) {
+		rc = errno;
+		ERROR("clock: allocating calibration structures, %s\n", strerror(rc));
+		goto finish;
+        }
+
+	/* Record current offsets */
+	for (i = 0; i < count; i++) {
+		struct sfptpd_clock *clock = snapshot[i];
+		struct sfptpd_clock_calibration *calibration = calibrations + i;
+
+		rc = sfptpd_interface_hw_timestamping_suspend(&calibration->intf_query,
+							      clock->u.nic.device_idx);
+		if (rc != 0)
+			calibration->error = true;
+		rc = sfptpd_clock_compare(clock, sys, &calibration->offset);
+		if (rc != 0) {
+			ERROR("clock %s: dedup1, %s\n",
+			      clock->short_name, strerror(rc));
+			calibration->error = true;
+		}
+	}
+
+	/* Check effect of adjusting freq for each clock */
+	for (i = 0; i < count; i++) {
+		struct sfptpd_clock *clock = snapshot[i];
+		struct sfptpd_clock_calibration *calibration = calibrations + i;
+		struct sfptpd_timespec mod_offset;
+
+		if (calibration->error)
+			continue;
+
+		/* Apply temporary offset */
+		sfptpd_time_add(&mod_offset, &calibration->offset, &test_adjustment);
+		rc = sfptpd_clock_adjust_time(clock, &test_adjustment);
+		if (rc != 0) {
+			ERROR("clock %s: dedup2, %s\n",
+			      clock->short_name, strerror(rc));
+			calibration->error = true;
+		}
+		usleep(settle);
+
+		/* Check new offset */
+		rc = sfptpd_clock_compare(clock, sys, &t);
+		if (rc != 0) {
+			ERROR("clock %s: dedup3, %s\n",
+			      clock->short_name, strerror(rc));
+			calibration->error = true;
+		}
+		if (!sfptpd_time_equal_within(&t, &mod_offset, &threshold)) {
+			INFO("clock %s: deduplication: did not take test adjustment\n",
+			     clock->short_name);
+			calibration->inert = true;
+		}
+
+		/* Check if offset impacts other clocks */
+		for (j = 0; j < count; j++) {
+			struct sfptpd_clock *other = snapshot[j];
+			if (i == j)
+				continue;
+			if (calibrations[j].error)
+				continue;
+
+			rc = sfptpd_clock_compare(other, sys, &t);
+			if (rc != 0) {
+				ERROR("clock %s: dedup4, %s\n",
+				      other->short_name, strerror(rc));
+				calibrations[j].error = true;
+			}
+			if (!sfptpd_time_equal_within(&t, &calibrations[j].offset, &threshold)) {
+				struct sfptpd_timespec mod2;
+				TRACE_L4("clock %s: clock %s: has changed\n",
```

Comment:
nit - maybe update this. It means that while adjusting clock1, clock2 has varied. (But we haven't fully confirmed that they are linked.)